src/Data/BitVector/Sized.hs

```diff
@@ -9,47 +9,100 @@
+'Data.BitVector.Sized.Signed.SignedBV' or
+'Data.BitVector.Sized.Unsigned.UnsignedBV'.
+
+This module should be imported qualified, as many of the names clash
+with those in Prelude or other base packages.
 -}
 
 module Data.BitVector.Sized
-  ( -- * BitVector type
-    BitVector, pattern BitVector
-  , bitVector, bitVector'
-  , bv0
-    -- * Bitwise operations (width-preserving)
-    -- | These are alternative versions of some of the 'Data.Bits' functions where we
-    -- do not need to know the width at compile time. They are all width-preserving.
-  , bvAnd, bvOr, bvXor
-  , bvComplement
-  , bvShift, bvShiftL, bvShiftRA, bvShiftRL, bvRotate
-  , bvWidth
-  , bvTestBit
-  , bvPopCount
-  , bvTruncBits
+  ( -- * 'BV.BV' type
+    BV.BV, pattern BV
+    -- * 'NatRepr's (from parameterized-utils)
+  , NatRepr
+  , knownNat
+    -- * Constructors
+  , mkBV, mkBVUnsigned, mkBVSigned
+  , unsignedClamp, signedClamp
+  , minUnsigned, maxUnsigned
+  , minSigned, maxSigned
+    -- * Construction from @Word@ and @Int@
+  , word8, word16, word32, word64
+  , int8, int16, int32, int64
+  , bytesBE, bytesLE
+    -- * Conversions to Integer and Natural
+  , asSigned
+  , asUnsigned
+  , asNatural
+    -- * Bits operations (width-preserving)
+    -- | 'BV' versions of the functions in @Data.Bits@.
+  , and, or, xor
+  , complement
+  , shl, lshr, ashr, rotateL, rotateR
+  , zero, one, width
+  , bit, bit'
+  , setBit, setBit'
+  , clearBit, clearBit'
+  , complementBit, complementBit'
+  , testBit, testBit'
+  , popCount
+  , ctz, clz
+  , truncBits
     -- * Arithmetic operations (width-preserving)
-  , bvAdd, bvMul
-  , bvQuotU, bvQuotS
-  , bvRemU, bvRemS
-  , bvAbs, bvNegate
-  , bvSignum
-  , bvLTS, bvLTU
+  , add, sub, mul
+  , uquot, squot, sdiv
+  , urem, srem, smod
+  , uquotRem, squotRem, sdivMod
+  , abs, negate
+  , signBit
+  , signum
+  , slt, sle, ult, ule
+  , umin, umax
+  , smin, smax
     -- * Variable-width operations
     -- | These are functions that involve bit vectors of different lengths.
-  , bvConcat, (<:>), bvConcatMany, bvConcatMany'
-  , bvExtract, bvExtract'
-  , bvZext, bvZext'
-  , bvSext, bvSext'
-    -- * Conversions to Integer
-  , bvIntegerU
-  , bvIntegerS
-    -- * Byte decomposition
-  , bvGetBytesU
+  , concat
+  , select, select'
+  , zext
+  , sext
+  , trunc, trunc'
+  , mulWide
+    -- * Enum operations
+  , succUnsigned, succSigned
+  , predUnsigned, predSigned
+  , enumFromToUnsigned, enumFromToSigned
+    -- * Pretty printing
+  , ppHex
+  , ppBin
+  , ppOct
+  , ppDec
   ) where
 
-import Data.BitVector.Sized.Internal
+import Data.BitVector.Sized.Internal hiding (BV(..))
+import qualified Data.BitVector.Sized.Internal as BV
+
+import Data.Parameterized.NatRepr (knownNat, NatRepr)
+import Prelude (Integer)
+
+-- | Get the underlying 'Integer' representation from a 'BV'. We
+-- guarantee that @(\(BV.BV x) -> x) == BV.toUnsigned@.
+pattern BV :: Integer -> BV.BV w
+pattern BV x <- BV.BV x
+{-# COMPLETE BV #-}
```

Comment:
Is this pattern just here for future backwards-compatibility?

Reply:
No -- the point is to allow pattern matches, but not export the constructor. If there's a cleaner way to do this, I'm all for it.

Reply:
What are the use cases for pulling out the `Integer` without going through the signed or unsigned accessors?  I'm curious - not opposed, but it seems ideal to keep the representation hidden entirely if possible

Reply:
It's really just an alternative to `asUnsigned`. Might make sense to call it `UBV` or something, so you know you're not actually using a constructor.